### PR TITLE
Equation 27: Determinant of 4x4 Identity + some matrix

### PR DESCRIPTION
**Pull request by MohanadAhmed**
# Equation 27: Determinant of 4x4 Identity + some matrix
The field of the matrix should be characteristic zero (includes integers If I understand correctly) since we need numbers like 3 6 and division by these numbers.

$$ \begin{aligned}
\text{det}(\mathbf{I}+\mathbf{A})= & 1+\text{det}(\mathbf{A})+\text{Tr}(\mathbf{A}) \\ 
& + \frac{1}{2} \left[ \text{Tr}(\mathbf{A})^2-\frac{1}{2} \text{Tr}\(\mathbf{A}^2 \right] \\
& + \frac{1}{6} \left[ \text{Tr}(\mathbf{A})^3- 3\text{Tr}(\mathbf{A}) \text{Tr}\mathbf{A}^2+2 \text{Tr}(\mathbf{A}^3)\right]
\end{aligned}$$

The file `mat_eq27_lib.lean` contains the bulk of the rather ugly brute force proof. This includes lemmas:

1. `trace_a_fin4`: $\text{Tr}[A] = A_{00} + A_{11} + A_{22} + A_{33}$
2. `trace_a_squared`, 
$$\text{Tr}[A^2] = A_{00}^2 + A_{11}^2 + A_{22}^2 + A_{33}^2 + 2A_{01}A_{10} + 2A_{02}A_{20} + 2A_{03}A_{30} +2A_{12}A_{21} + 2A_{13}A_{31} + 2A_{23}A_{32}$$
3. `trace_a_cubed`: $\text{Tr}[A^3] = $, 
$A_{00} (A_{00}^2 + A_{01} A_{10} + A_{02} A_{20} + A_{03} A_{30}) +     $, 
$A_{11} (A_{11}^2 + A_{01} A_{10} + A_{12} A_{21} + A_{13} A_{31}) +     $, 
$A_{22} (A_{22}^2 + A_{02} A_{20} + A_{12} A_{21} + A_{23} A_{32}) +     $, 
$A_{33} (A_{33}^2 + A_{03} A_{30} + A_{13} A_{31} + A_{23} A_{32}) +     $, 
$A_{10} (A_{00} A_{01} + A_{01} A_{11} + A_{02} A_{21} + A_{03} A_{31}) +$, 
$A_{20} (A_{00} A_{02} + A_{01} A_{12} + A_{02} A_{22} + A_{03} A_{32}) +$, 
$A_{01} (A_{00} A_{10} + A_{10} A_{11} + A_{12} A_{20} + A_{13} A_{30}) +$, 
$A_{30} (A_{00} A_{03} + A_{01} A_{13} + A_{02} A_{23} + A_{03} A_{33}) +$, 
$A_{21} (A_{02} A_{10} + A_{11} A_{12} + A_{12} A_{22} + A_{13} A_{32}) +$, 
$A_{02} (A_{00} A_{20} + A_{10} A_{21} + A_{20} A_{22} + A_{23} A_{30}) +$, 
$A_{31} (A_{03} A_{10} + A_{11} A_{13} + A_{12} A_{23} + A_{13} A_{33}) +$, 
$A_{12} (A_{01} A_{20} + A_{11} A_{21} + A_{21} A_{22} + A_{23} A_{31}) +$, 
$A_{03} (A_{00} A_{30} + A_{10} A_{31} + A_{20} A_{32} + A_{30} A_{33}) +$, 
$A_{32} (A_{03} A_{20} + A_{13} A_{21} + A_{22} A_{23} + A_{23} A_{33}) +$, 
$A_{13} (A_{01} A_{30} + A_{11} A_{31} + A_{21} A_{32} + A_{31} A_{33}) +$, 
$A_{23} (A_{02} A_{30} + A_{12} A_{31} + A_{22} A_{32} + A_{32} A_{33}) $

4. `det4': determinant of a 4x4 matrix in explicit form. $\text{det}[A] = $
$A_{00}A_{11}A_{22}A_{33} - A_{00}A_{11}A_{23}A_{32} - A_{00}A_{12}A_{21}A_{33} +$, 
$A_{00}A_{12}A_{23}A_{31} + A_{00}A_{13}A_{21}A_{32} - A_{00}A_{13}A_{22}A_{31} -$,
$A_{01}A_{10}A_{22}A_{33} + A_{01}A_{10}A_{23}A_{32} + A_{01}A_{12}A_{20}A_{33} -$, 
$A_{01}A_{12}A_{23}A_{30} - A_{01}A_{13}A_{20}A_{32} + A_{01}A_{13}A_{22}A_{30} +$,
$A_{02}A_{10}A_{21}A_{33} - A_{02}A_{10}A_{23}A_{31} - A_{02}A_{11}A_{20}A_{33} +$, 
$A_{02}A_{11}A_{23}A_{30} + A_{02}A_{13}A_{20}A_{31} - A_{02}A_{13}A_{21}A_{30} -$,
$A_{03}A_{10}A_{21}A_{32} + A_{03}A_{10}A_{22}A_{31} + A_{03}A_{11}A_{20}A_{32} -$, 
$A_{03}A_{11}A_{22}A_{30} - A_{03}A_{12}A_{20}A_{31} + A_{03}A_{12}A_{21}A_{30}$

5. `det4_one_add_a`: $\text{det}(I + A)$
6. `trace_a_all_squared`: (trace A) * (trace A)
7. `eq_27_rhs_part1`, `eq_27_rhs_part2`, `eq_27_rhs`, `eq_27_before_last`: lemmas to help break down the ugly computation and avoid deterministic timeouts.

### CODE REVIEW

**Comment by eric-wieser** [comment]
Wow, that is rather brute force!

I wonder if it's easy to prove [the more general Girard-Waring formula](https://math.stackexchange.com/a/2779104/1896), which then recovers these as special cases.

**Comment by MohanadAhmed** [comment]
> Wow, that is rather brute force!
> 
> I wonder if it's easy to prove [the more general Girard-Waring formula](https://math.stackexchange.com/a/2779104/1896), which then recovers these as special cases.

Looks ugly:

$$ \Delta_A(\lambda)=\det(\lambda+A)=\sum_{r=0}^{n}P_r(A)\lambda^r $$
with 
$$P_r(A)=\sum_{\pi\in\Pi(n-r)}(-1)^{\|\pi\|-n+r}\prod_{m=1}^{n-r}\frac{(Tr(A^m))^{\pi_m}}{m^{\pi_m}\pi_m!}$$

I don't think it will be easier than the brute force solution. But then my lean skills are still just rewrites!!

**Comment by MohanadAhmed** [comment]
Do you think you can merge this one? I think it is quite limited compared to the other ones

File: src/matrix_cookbook/mat_eq27_lib.lean [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    2*A 1 2*A 2 1 + 2*A 1 3*A 3 1 + 2*A 2 3*A 3 2 := 
+begin
+  rw matrix.trace,
+  simp only [diag],
+  rw pow_two, rw mul_eq_mul, rw fin.sum_univ_four,
+  repeat{rw mul_apply, rw fin.sum_univ_four},
+  ring,
+end
+
+lemma trace_a_cubed {A : matrix (fin 4) (fin 4) R}:
```
Comment (eric-wieser) [inline src/matrix_cookbook/mat_eq27_lib.lean:34]: Generally lemma names shouldn't refer to the names of the variables they use; so this one would be

```suggestion
lemma trace_pow_three_fin_four {A : matrix (fin 4) (fin 4) R}:
```

File: src/matrix_cookbook/mat_eq27_lib.lean [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    A 2 3*(A 0 2*A 3 0 + A 1 2*A 3 1 + A 2 2*A 3 2 + A 3 2*A 3 3)
+ := 
+begin
+  rw pow_three, rw matrix.trace,
+  rw fin.sum_univ_four, rw mul_eq_mul, rw mul_eq_mul, simp only[diag],
+  repeat{rw mul_apply, rw fin.sum_univ_four},
+  ring,
+end
+
+lemma det4 {A : matrix (fin 4) (fin 4) R}:
```
Comment (eric-wieser) [inline src/matrix_cookbook/mat_eq27_lib.lean:60]:
```suggestion
lemma det_fin_four (A : matrix (fin 4) (fin 4) R) :
```

File: src/matrix_cookbook/mat_eq27_lib.lean [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+  have : (2: fin 4).succ_above 0 = 0, {refl,},  rw this,
+  have : (2: fin 4).succ_above 1 = 1, {refl,},  rw this,
+  have : (2: fin 4).succ_above 2 = 3, {refl,},  rw this,
+  have : (3: fin 4).succ_above 0 = 0, {refl,},  rw this,
+  have : (3: fin 4).succ_above 1 = 1, {refl,},  rw this,
+  have : (3: fin 4).succ_above 2 = 2, {refl,},  rw this,
+  ring,
+end
+
+lemma det4_one_add_a {A : matrix (fin 4) (fin 4) R}:
```
Comment (eric-wieser) [inline src/matrix_cookbook/mat_eq27_lib.lean:107]:
```suggestion
lemma det_one_add_fin_four (A : matrix (fin 4) (fin 4) R) :
```

File: src/matrix_cookbook/mat_eq27_lib.lean [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+  A 1 1*A 3 3 - A 1 3*A 3 1 + A 2 2*A 3 3 - A 2 3*A 3 2 + 
+  A 0 0*A 1 1*A 2 2 - A 0 0*A 1 2*A 2 1 - A 0 1*A 1 0*A 2 2 + A 0 1*A 1 2*A 2 0 + A 0 2*A 1 0*A 2 1 - A 0 2*A 1 1*A 2 0 + A 0 0*A 1 1*A 3 3 - A 0 0*A 1 3*A 3 1 - 
+  A 0 1*A 1 0*A 3 3 + A 0 1*A 1 3*A 3 0 + A 0 3*A 1 0*A 3 1 - A 0 3*A 1 1*A 3 0 + A 0 0*A 2 2*A 3 3 - A 0 0*A 2 3*A 3 2 - A 0 2*A 2 0*A 3 3 + A 0 2*A 2 3*A 3 0 + 
+  A 0 3*A 2 0*A 3 2 - A 0 3*A 2 2*A 3 0 + A 1 1*A 2 2*A 3 3 - A 1 1*A 2 3*A 3 2 - A 1 2*A 2 1*A 3 3 + A 1 2*A 2 3*A 3 1 + A 1 3*A 2 1*A 3 2 - A 1 3*A 2 2*A 3 1
+:= 
+begin
+  rw det4_one_add_a, rw det4, ring,
+end
+
+lemma trace_a_all_squared {A : matrix (fin 4) (fin 4) R}:
```
Comment (eric-wieser) [inline src/matrix_cookbook/mat_eq27_lib.lean:150]:
```suggestion
lemma sq_trace_fin_four({A : matrix (fin 4) (fin 4) R) :
```

File: src/matrix_cookbook/mat_eq27_lib.lean [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+  A 0 3*A 2 0*A 3 2 - A 0 3*A 2 2*A 3 0 + A 1 1*A 2 2*A 3 3 - A 1 1*A 2 3*A 3 2 - A 1 2*A 2 1*A 3 3 + A 1 2*A 2 3*A 3 1 + A 1 3*A 2 1*A 3 2 - A 1 3*A 2 2*A 3 1
+:= 
+begin
+  have h2: (2:R) ≠ 0, {norm_num,},
+  have h6: (6:R) ≠ 0, {norm_num,},
+  rw eq_27_rhs_part1, rw ← mul_assoc, rw one_div_mul_cancel h6, 
+  rw one_mul,
+  rw eq_27_rhs_part2, rw ← mul_assoc, rw one_div_mul_cancel h2, 
+  rw one_mul,
+  rw trace_a_fin4, ring,
```
Comment (eric-wieser) [inline src/matrix_cookbook/mat_eq27_lib.lean:194]: Can you change to one tactic per line (here and throughout)? So

```suggestion
  rw eq_27_rhs_part1,
  rw ← mul_assoc,
  rw one_div_mul_cancel h6, 
  rw one_mul,
  rw eq_27_rhs_part2,
  rw ← mul_assoc, rw one_div_mul_cancel h2, 
  rw one_mul,
  rw trace_a_fin4, ring,
```
or
```suggestion
  rw [eq_27_rhs_part1, ← mul_assoc, rw one_div_mul_cancel h6, 
    one_mul, eq_27_rhs_part2, ← mul_assoc, one_div_mul_cancel h2, one_mul,
    trace_a_fin4],
  ring,
```

File: src/matrix_cookbook/mat_eq27_lib.lean [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+
+
+lemma eq_27_before_last {A : matrix (fin 4) (fin 4) R} :
+  det (1 + A) - det A - 1 = trace A + 
+    (1/2)*( (trace A)^2 - trace (A^2)) + 
+    (1/6)*( (trace A)^3 - 3*trace A * trace (A^2) + 2 * trace (A^3) ) := 
+begin
+  rw eq_27_lhs, rw eq_27_rhs,
+end
```
Comment (eric-wieser) [inline src/matrix_cookbook/mat_eq27_lib.lean:205]: You could combine these two into a single proof as something like
```suggestion
begin
  transitivity  A 0 0 + A 1 1 + A 2 2 + A 3 3 + 
  A 0 0*A 1 1 - A 0 1*A 1 0 + A 0 0*A 2 2 - A 0 2*A 2 0 + A 0 0*A 3 3 - A 0 3*A 3 0 + A 1 1*A 2 2 - A 1 2*A 2 1 + 
  A 1 1*A 3 3 - A 1 3*A 3 1 + A 2 2*A 3 3 - A 2 3*A 3 2 + 
  A 0 0*A 1 1*A 2 2 - A 0 0*A 1 2*A 2 1 - A 0 1*A 1 0*A 2 2 + A 0 1*A 1 2*A 2 0 + A 0 2*A 1 0*A 2 1 - A 0 2*A 1 1*A 2 0 + A 0 0*A 1 1*A 3 3 - A 0 0*A 1 3*A 3 1 - 
  A 0 1*A 1 0*A 3 3 + A 0 1*A 1 3*A 3 0 + A 0 3*A 1 0*A 3 1 - A 0 3*A 1 1*A 3 0 + A 0 0*A 2 2*A 3 3 - A 0 0*A 2 3*A 3 2 - A 0 2*A 2 0*A 3 3 + A 0 2*A 2 3*A 3 0 + 
  A 0 3*A 2 0*A 3 2 - A 0 3*A 2 2*A 3 0 + A 1 1*A 2 2*A 3 3 - A 1 1*A 2 3*A 3 2 - A 1 2*A 2 1*A 3 3 + A 1 2*A 2 3*A 3 1 + A 1 3*A 2 1*A 3 2 - A 1 3*A 2 2*A 3 1,
  { -- put your lhs proof here
    sorry },
  { -- put your rhs proof here,
    sorry },
```

Comment (MohanadAhmed) [reply src/matrix_cookbook/mat_eq27_lib.lean:204]: You mean get rid of the lemmas eq_27_lhs and eq_27_rhs and put them all in the before last lemma?

Comment (eric-wieser) [reply src/matrix_cookbook/mat_eq27_lib.lean:204]: Yes, exactly!

**Comment by eric-wieser** [comment]
Can you `git fetch; git merge origin/master`? I just fixed the github configuration so that it runs your code and checks it works!

**Comment by eric-wieser** [comment]
That button is for CI, it's to stop random users mining cryptocurrency on my CI quota! I've hit the button now.

**Comment by eric-wieser** [comment]
Can you merge master one more time?

File: src/matrix_cookbook/mat_eq27_lib.lean [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    rw pow_zero,
+  },
+  have a3 : (-1:R)^(↑(3: fin 4)) = -1, {
+    rw neg_one_pow_eq_pow_mod_two, 
+    have : ↑(3: fin 4) = (3:ℕ), {refl,}, 
+    rw this, 
+    have : 3 % 2  = 1, by refl, 
+    rw this, 
+    rw pow_one,
+  },
```
Comment (eric-wieser) [inline src/matrix_cookbook/mat_eq27_lib.lean:93]: These CI errors are telling you to format things as
```suggestion
  have a3 : (-1:R)^(↑(3: fin 4)) = -1,
  { rw neg_one_pow_eq_pow_mod_two, 
    have : ↑(3: fin 4) = (3:ℕ), {refl,}, 
    rw this, 
    have : 3 % 2  = 1, by refl, 
    rw this, 
    rw pow_one },
```
which is the mathlib style

Comment (MohanadAhmed) [reply src/matrix_cookbook/mat_eq27_lib.lean:93]: Which part is in the error message is saying that?

```
Run ! (_target/deps/mathlib/scripts/lint-style.py $(find src -name '*.lean') || true) | (grep -vE 'ERR_COP|ERR_LIN' || true) | grep '^'
Error: ERR_MOD: Module docstring missing, or too late
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Warning: WRN_BRC: Probable misformatting of curly braces
Error: Process completed with exit code 1.
```

What is ERR_MOD??

Comment (MohanadAhmed) [reply src/matrix_cookbook/mat_eq27_lib.lean:93]: Also can I run the linter locally? Is there an extension or some cmd tool?

Comment (MohanadAhmed) [reply src/matrix_cookbook/mat_eq27_lib.lean:93]: Never mind found it:

```
python .\_target\deps\mathlib\scripts\lint-style.py .\src\matrix_cookbook\1_basic.lean
::error file=src\matrix_cookbook\1_basic.lean,line=27,code=ERR_LIN::ERR_LIN: Line has more than 100 characters
::error file=src\matrix_cookbook\1_basic.lean,line=31,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=34,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=35,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=43,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=47,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=53,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=58,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=68,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=93,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=99,code=ERR_LIN::ERR_LIN: Line has more than 100 characters   
::error file=src\matrix_cookbook\1_basic.lean,line=1,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=2,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=3,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=4,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=5,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=6,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=7,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=0,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=0,code=ERR_COP::ERR_COP: Malformed or missing copyright header 
::error file=src\matrix_cookbook\1_basic.lean,line=10,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\1_basic.lean,line=0,code=ERR_COP::ERR_COP: Malformed or missing copyright header 
::error file=src\matrix_cookbook\1_basic.lean,line=12,code=ERR_COP::ERR_COP: Malformed or missing copyright header

python .\_target\deps\mathlib\scripts\lint-style.py .\src\matrix_cookbook\mat_eq27_lib.lean
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=121,code=ERR_LIN::ERR_LIN: Line has more than 100 characters
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=123,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=124,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=125,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=126,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=127,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=128,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=129,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=162,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=163,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=164,code=ERR_LIN::ERR_LIN: Line has more than 100 characters  
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=1,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=2,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=3,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=4,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=5,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=6,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=7,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=8,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=0,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=10,code=ERR_COP::ERR_COP: Malformed or missing copyright header
::error file=src\matrix_cookbook\mat_eq27_lib.lean,line=10,code=ERR_MOD::ERR_MOD: Module docstring missing, or too late
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=80,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=85,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=86,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=93,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=96,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=100,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=200,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=204,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=205,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
::warning file=src\matrix_cookbook\mat_eq27_lib.lean,line=212,code=WRN_BRC::WRN_BRC: Probable misformatting of curly braces
```

Comment (eric-wieser) [reply src/matrix_cookbook/mat_eq27_lib.lean]: Note that in CI I'm ignoring `ERR_COP` and `ERR_LIN`

File: src/matrix_cookbook/mat_eq27_lib.lean [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+requirement, which I satsified by requiring a field with characteristic zero.
+
+-/
+
+open_locale matrix big_operators
+open matrix
+
+variables {R : Type*}[field R][char_zero R]
+
+lemma trace_fin_four {A : matrix (fin 4) (fin 4) R} :
```
Comment (eric-wieser) [inline src/matrix_cookbook/mat_eq27_lib.lean:33]: Can you move this lemma and `det_fin_four` to `for_mathlib/data/matrix.lean`?

Comment (eric-wieser) [reply src/matrix_cookbook/mat_eq27_lib.lean:33]: (the other lemmas are rather specialized, but I think these do belong in mathlib)

**Comment by MohanadAhmed** [comment]
One more press @eric-wieser :)

**Comment by eric-wieser** [comment]
Once your first PR is merged I won't have to push the button any more.

File: src/matrix_cookbook/mat_eq27_lib.lean [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    A 1 2*A 2 3*A 3 1 + A 1 3*A 2 1*A 3 2 - A 1 3*A 2 2*A 3 1, 
+  { rw [det_one_add_fin_four, det_fin_four], 
+    ring, },
+  { have h2 : (2:R) ≠ 0, { norm_num },
+    have h6 : (6:R) ≠ 0, { norm_num },
+    rw [eq_27_rhs_part1, ← mul_assoc, one_div_mul_cancel h6, one_mul], 
+    rw [eq_27_rhs_part2, ← mul_assoc, one_div_mul_cancel h2, one_mul],
+    rw trace_fin_four,
+    ring, },
+end
```
Comment (eric-wieser) [inline src/matrix_cookbook/mat_eq27_lib.lean:150]: Unless you think the algebraic expansion is particularly illuminative,
```suggestion
lemma eq_27_before_last {R} [field R] [char_zero R] (A : matrix (fin 4) (fin 4) R)  :
  det (1 + A) - det A - 1 = trace A + 
    (1/2)*( (trace A)^2 - trace (A^2)) + 
    (1/6)*( (trace A)^3 - 3*trace A * trace (A^2) + 2 * trace (A^3) ) := 
begin
  field_simp,
  rw [det_one_add_fin_four, det_fin_four, trace_pow_three_fin_four, trace_pow_two_fin_four,
    sq_trace_fin_four, trace_fin_four],
  ring,
end
```

**Comment by eric-wieser** [comment]
> Do you think you can merge this one? I think it is quite limited compared to the other ones

Yes, I think this is looking close to mergeable :). Thanks for isolating this into small chunk.
I'd certainly like to see the Girard-Waring version, but I'd be happy with a TODO comment about it and merging the brute-force version for now.

I think the only thing I'm not sure of right now is where `mat_eq27_lib.lean` should live. Perhaps `for_mathlib/matrix/girard_waring.lean`, with a note that it only actually contains one very contrived special case?

**Comment by MohanadAhmed** [comment]
Anymore barriers to merge?

**Comment by eric-wieser** [comment]
I've made a new `lib` folder. Thanks for that final golf, I was going to suggest removing `eq27_before_last` anyway!

**Comment by eric-wieser** [comment]
Thanks :tada:

**Comment by MohanadAhmed** [comment]
Thanks! That was a long one! See you on the next one

**Pull request merged**